### DISCUSSION TITLE
As a way of extending the support for data migration I have added some code

### DISCUSSION THREAD

**Discussion by juandent** [fnc12/sqlite_orm#995] [Show and tell]
The order in which tables can be dropped is subject to certain restrictions: tables with dependent rows cannot be dropped.
To be able to drop tables with the certainty that we will drop them in correct order, we need an algorithm that "sorts" the types. I have implemented this algorithm. The idea is to use it like this:

```C++
		SQLCookbookDb db;
		auto vec0 = db.drop_table<0>(storage);
		auto vec1 = db.drop_table<1>(storage);
		auto vec2 = db.drop_table<2>(storage);
		auto vec3 = db.drop_table<3>(storage);
		auto vec4 = db.drop_table<4>(storage);

		storage.sync_schema();                          // dropping the tables allows for schema changes to be synchronized
                                                                              // as was carefully analyzed in the Guide which I have written

		db.replace_table(storage, vec4);
		db.replace_table(storage, vec3);
		db.replace_table(storage, vec2);
		db.replace_table(storage, vec1);
		db.replace_table(storage, vec0);
```
This behavior works like a stack and it would be nice to create a vector of vectors but how since each table represents a different type - so vec0 may be something like ```std::vector<Album>``` and vec1 something like ```std::vector<Artist>```, etc, clearly impossible to create a stack of vectors... if there is a way to make this more generic please let me know... I don't like to hang my trust to the naming of the vectors (i.e. vec0, vec1,...). If there is a way please let me know!

I am using the CRTP to define a general Database template class which has a couple of extension points:

```C++
template<typename ConcreteDatabase>
struct Database
{
	Database()
	{
	}
	auto getDropOrder() const noexcept
	{
		return static_cast<const ConcreteDatabase*>(this)->getDropOrderImpl();
	}
	auto getInsertOrder() const noexcept
	{
		auto dropOrder = getDropOrder();
		return TableOrder::tuple_reverse<decltype(dropOrder)>::type();
	}
	template<size_t Pos, typename Storage>
	auto drop_table(Storage storage)
	{
		auto drop_order = getDropOrder();
		using Element = std::tuple_element_t<Pos, decltype(drop_order)>;
		std::vector<Element> vec;
		vec = storage.get_all<Element>();
		storage.drop_table(storage.tablename<Element>());
		return vec;
	}
	template<typename Storage, typename Vector>
	void replace_table(Storage storage, const Vector& vec)
	{
		storage.replace_range(vec.begin(), vec.end());
	}
	size_t getCountTables() const noexcept
	{
		return static_cast<const ConcreteDatabase*>(this)->getCountTablesImpl();
	}
};
```  

which is common for all projects. A concrete class for each storage schema then follows:

```C++
struct SQLCookbookDb : public Database<SQLCookbookDb>
{
	using ListOfTables = std::tuple<Artist, Department, Album, Employee, EmpBonus>;
	auto getDropOrderImpl() const noexcept
	{
		auto newTuple = TableOrder::tuple_divide<Employee, Department, ListOfTables>::convert();
		auto newerTuple = TableOrder::tuple_divide<EmpBonus, Employee, decltype(newTuple)>::convert();
		auto newestTuple = TableOrder::tuple_divide<Album, Artist, decltype(newerTuple)>::convert();
		return newestTuple;
	}
	size_t getCountTablesImpl() const noexcept
	{
		return std::tuple_size_v<ListOfTables>;
	}
};
```

the name tuple_derived should be replaced by something more appropriate - it is a temporary name. The implementation of the ordering algorithm is as follows (this needs to be called for every pair of types that require some order between them):

```C++
	template<typename I, typename D, typename Tuple>
	struct tuple_divide
	{
	private:
		static constexpr int pos_i = find_in_tuple<I, 0, size, Tuple>::pos;
		static constexpr int pos_d = find_in_tuple<D, 0, size, Tuple>::pos;
                    // ......
	public:
		constexpr static auto convert()               // algorithm
		{
			Tuple t{};

			if constexpr (pos_i > pos_d)						// relative position of Tables is incorrect
			{
				auto t3 = tail<pos_i + 1>(t);					// get tail after i

				auto t2plus = tail<pos_d + 1>(t);				// get tail after d

				auto t2 = head<pos_i - pos_d - 1>(t2plus);		// get elements between d and i

				auto t1 = head<pos_d>(t);						// get elements before d

				auto pair = std::tuple<I, D>();					// create the insertion tuple

				auto mix = std::tuple_cat(t1, pair, t2, t3);	// concatenate tuples into final mix

				return mix;
			}
			else												// no changes needed, return Tuple instance
			{
				return t;
			}
		}
```

I would like to know what you think about this...
I will prepare .h files to package this physically. I suppose I can then upload the files for your review.

Thanks in advance,
Juan Dent

**Comment by juandent** [comment]
BTW, I am not suggesting that we drop all the tables on every execution of each of our applications. Only when there have been changes that are not updated (because sync_schema cannot detect a change for instance) or in cases where sync_schema would drop and loose data (see my Guide, data migration)...

**Comment by juandent** [comment]
I would like to upload the code files of the above mentioned extensions... what is the best way to do this?

**Reply by fnc12** [reply]
PR

**Comment by fnc12** [comment]
Actually there can be a schema where one cannot detect a strict sort order of tables cause table A can have a FK to table B and reverse simultaneously. Example: table `books` has a FK to `category` storing book's category and table `category` can has FK to `books` with the latest added book of this category. Such a schema can be created using external SQLite client, cannot be created using `sync_schema` but can be supported by `sqlite_orm`. Developer may want to alter such a schema with `sqlite_orm`. How would you do this?

**Comment by juandent** [comment]
Sqlite studio uses pragma foreign_keys 0 before dropping and creating tables and then resets it to 1
This is the reason why I think we should have access to this pragma
Sync_schema would be run by shutting down foreign_keys to 0 doing the work, and then resetting to 1

Actually foreign_keys should be 0 when we start dropping tables and reset to 1 after all data is reloaded

This should be done with RAII IDIOM to ensure the pragma is reset!

That is the way one can create the schema in a SQL client but can also be created within the full data migration using sync_schema

Cheers 🍻 
Juan



Sent from my iPhone
s to thi1

**Comment by juandent** [comment]
Since all inserting/updating is done with PRAGMA FOREIGN_KEY 1, we know the DB is in a valid referential integrity state before the sync_schema(), then turn off PRAGMA FOREIGN KEY, load the tables, drop them, reset the schema and reinsert the tables, resetting FOREIGN_KEY to 1
This is the only place where we touch this PRAGMA...

**Reply by juandent** [reply]
this is how SQLITESTUDIO does the schema changes!!

**Reply by juandent** [reply]
For instance, I added a column to an existing table and this is the query to be executed in sqlitestudio:

```SQL
PRAGMA foreign_keys = 0;

CREATE TABLE sqlitestudio_temp_table AS SELECT *
                                          FROM employees_bk;

DROP TABLE employees_bk;

CREATE TABLE employees_bk (
    EmployeeId INTEGER       PRIMARY KEY AUTOINCREMENT
                             NOT NULL,
    LastName   NVARCHAR (20) NOT NULL,
    FirstName  NVARCHAR (20) NOT NULL,
    Title      NVARCHAR (30),
    ReportsTo  INTEGER,
    BirthDate  DATETIME,
    HireDate   DATETIME,
    Address    NVARCHAR (70),
    City       NVARCHAR (40),
    State      NVARCHAR (40),
    Country    NVARCHAR (40),
    PostalCode NVARCHAR (10),
    Phone      NVARCHAR (24),
    Fax        NVARCHAR (24),
    Email      NVARCHAR (60),
    country    TEXT          DEFAULT unknown,
    FOREIGN KEY (
        ReportsTo
    )
    REFERENCES employees_bk (EmployeeId) ON DELETE NO ACTION
                                         ON UPDATE NO ACTION
);

INSERT INTO employees_bk (
                             EmployeeId,
                             LastName,
                             FirstName,
                             Title,
                             ReportsTo,
                             BirthDate,
                             HireDate,
                             Address,
                             City,
                             State,
                             Country,
                             PostalCode,
                             Phone,
                             Fax,
                             Email
                         )
                         SELECT EmployeeId,
                                LastName,
                                FirstName,
                                Title,
                                ReportsTo,
                                BirthDate,
                                HireDate,
                                Address,
                                City,
                                State,
                                Country,
                                PostalCode,
                                Phone,
                                Fax,
                                Email
                           FROM sqlitestudio_temp_table;

DROP TABLE sqlitestudio_temp_table;

PRAGMA foreign_keys = 1;
```
See? no need to worry about foreign keys!!! Can drop any table by itself!! Thanks to foreign_keys == 0 during the operation

**Reply by fnc12** [reply]
Why sqlitestudio does this weird combination of statements instead of calling `ALTER TABLE ... ADD COLUMN...`?

**Reply by juandent** [reply]
I don't know why but it fully creates the table anew which might be what some schema changes require... its probably the safest route although not the more efficient ... don't know really, but the point is that foreign_keys are set to 0 while the saving the data, dropping, recreating the table and reloading all data

**Comment by juandent** [comment]
It is necessary to disable foreign_key checking else the insert will not work in the general case where the row with the reportsto FK is not inserted before the current row!!
How else?
BTW, this makes it unnecessary to drop the tables in any specific order!!

**Reply by juandent** [reply]
actually we need only drop the table that we want to update its schema!! we can ignore dependent rows!!

**Reply by fnc12** [reply]
ok I see. It makes sense. We can disable `PRAGME foreign_keys` in the beginning of `sync_schema` call and enable it in the end. What do you think?

**Reply by juandent** [reply]
no, I think that we can reset foreign_keys to 1 only after loading all the data!

**Reply by juandent** [reply]
just like sqlitestudio does!!!

**Reply by juandent** [reply]
sync_schema() does the table creation but then we need to store the data which must have any dependent rows existing
what if there is a self-foreign key and the rows are not entered in the right order? we would have an exception and we would loose the data in the table!

**Reply by juandent** [reply]
that's why we need FOREIGN_KEY = 0 until after the loading!!

**Reply by juandent** [reply]
or just a cyclic foreign key as in your example

**Reply by juandent** [reply]
which table should be added first?

**Reply by fnc12** [reply]
I'd skip allowing users to modify `PRAGMA foreign_keys`. I'd better create a more advanced way of migrations instead to make it user friendly

**Reply by juandent** [reply]
we should have a schema_update object that sets FOREIGN_KEY to 0 and in its destructor sets FOREIGN_KEY to 1
Then load and drop the tables, then call sync_schema(), then store the data from the inmemory vectors

**Reply by juandent** [reply]
I agree with you 100%on : I'd skip allowing users to modify PRAGMA foreign_keys

**Reply by juandent** [reply]
the solution is:
1- a very simple object called schema_update, which sets FOREING_KEY to 0 in ctor 
2- load the data to vectors and drop the tables
3- call sync_schema()
4- store the data back to the tables from the vectors without foreign_keys being checked
5 - the destructor of schema_update sets FOREING_KEY to 1

**Reply by fnc12** [reply]
`a very simple object called schema_update` is almost the same as `manual enabling and disabling PRAGMA foreign_keys`

**Reply by juandent** [reply]
this ensures the programmer is aware you are disabling FK testing but also insures that it is reset to 1 at the end of the object's lifetime!!

**Reply by juandent** [reply]
the problem is for severe schema changes or with cyclic dependencies!!

**Reply by fnc12** [reply]
I know how it works. It is still almost manual control but I am sure it can be automated. Better idea is to make more advanced migrations which can handle data migration. will disable `PRAGMA foreign_keys` in the beginning and enable in the end

**Reply by juandent** [reply]
just don't set FK testing in sync_schema() --- it should be 0 at the start but will be reset to 1 at the migration end

**Reply by fnc12** [reply]
this is what I already did in the latest branch but you told that 'it is not enough Eugene'

**Reply by juandent** [reply]
> It is still almost manual control but I am sure it can be automated. Better idea is to make more advanced migrations which can handle data migration. will disable PRAGMA foreign_keys in the beginning and enable in the end
Okay!!

**Reply by juandent** [reply]
no because it does not cover the loading of the data... the FKs are checked at the loading which can give us exceptions (interruptions)

**Reply by fnc12** [reply]
ok but why then you told that `just don't set FK testing in sync_schema() --- it should be 0 at the start but will be reset to 1 at the migration end`? One of your statements collides with another.

**Reply by juandent** [reply]
I am talking about manual loading of the data which sync_schema() does not do in the cases where it does not identify a change of schema or loses all the table's data as when adding a primary key --- see my guide on data migration

**Reply by juandent** [reply]
this is it, see carefully... loading of data does not necessarily take place inside of sync_schema(), that is why resetting at end of the method is too early cause it does not cover the loading of the data which might have FK breakings. Only at the END of the loading can we reset the FK support -- which is not in sync_schema()

**Reply by fnc12** [reply]
I understood it from the first time. I was talking about colliding statements cause it is difficult to get a straight topic from this conversation cause you write a lot of short distributed messages

**Reply by fnc12** [reply]
I understand. Let's make migrations API instead of making public available `PRAGMA foreign_keys`. Why do we both repeat our thoughts for three times?

**Reply by juandent** [reply]
okay sorry about that
no need to repeat ourselves

**Reply by juandent** [reply]
do you see migration API as something doable in the short term?

**Reply by fnc12** [reply]
Not sure. It looks like a huge feature with huge amount of test cases

**Reply by juandent** [reply]
that is what i was trying to avoid --- instead having one object to encapsulate the FK checking is so simple -- the programmer is responsible for its correct use

**Reply by fnc12** [reply]
The only argument is that your solutions is quick to implement? What I see is that the quickest is not always the best

**Reply by juandent** [reply]
my argument is that its not that dangerous and yes that it is quick to implement. With so many features on our table such as views, we hardly have the luxury to wait for perfect solutions.. I may be wrong, I will agree with whatever you think its best: you are the parent of the baby and it is noble to want it to be as perfect as possible

**Reply by fnc12** [reply]
if we merge a feature like one object to encapsulate the FK checking then we'll be unable to unmerge after the next release if we want because we must keep backward compatibility

**Comment by fnc12** [comment]
@juandent please check out `sync-schema-fk-modification` branch - it disables `PRAGMA foreign_keys` in the beginning of `sync_schema` and enables in the end

**Reply by juandent** [reply]
this is not enough Eugene!!

**Reply by fnc12** [reply]
what is your idea instead?

**Reply by juandent** [reply]
see above please

**Comment by fnc12** [comment]
Let's concentrate on advanced migrations API. It will solve both cyclic FKs problem and FK enabled/disabled problem
